axelor-stock/src/main/java/com/axelor/apps/stock/service/StoredProductsServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+    this.storeProductsRepository = storeProductsRepository;
+    this.massStockMoveRepository = massStockMoveRepository;
+    this.stockMoveRepository = stockMoveRepository;
+    this.stockMoveLineRepository = stockMoveLineRepository;
+    this.stockMoveService = stockMoveService;
+    this.stockMoveLineService = stockMoveLineService;
+  }
+
+  @Override
+  public StoredProducts createStoredProduct(
+      Product product,
+      TrackingNumber trackingNumber,
+      BigDecimal currentQty,
+      StockLocation toStockLocation,
+      BigDecimal storedQty,
+      StockMoveLine stockMoveLine,
+      MassStockMove massStockMove) {
+    StoredProducts storedProducts = new StoredProducts();
+    storedProducts.setStoredProduct(product);
+    storedProducts.setTrackingNumber(trackingNumber);
+    storedProducts.setCurrentQty(currentQty);
+    storedProducts.setUnit(product.getUnit());
+    storedProducts.setToStockLocation(toStockLocation);
+    storedProducts.setStoredQty(storedQty);
+    storedProducts.setStockMoveLine(stockMoveLine);
+    storedProducts.setMassStockMove(massStockMove);
+    return storedProducts;
+  }
+
+  @Override
+  @Transactional(rollbackOn = Exception.class)
+  public void createStockMoveAndStockMoveLine(StoredProducts storedProducts)
+      throws AxelorException {
+    BigDecimal storedQty =
+        storedProducts.getStoredQty() != null ? storedProducts.getStoredQty() : null;
+    BigDecimal currentQty =
+        storedProducts.getCurrentQty() != null ? storedProducts.getCurrentQty() : null;
+
+    if (storedQty.compareTo(currentQty) == -1 && storedQty.compareTo(BigDecimal.ZERO) != 0) {
+      Product duplicateProduct = storedProducts.getStoredProduct();
+      TrackingNumber duplicateTrackingNumber = storedProducts.getTrackingNumber();
+      StockLocation duplicateToStockLocation = storedProducts.getToStockLocation();
+      StockMoveLine duplicateStockMoveLine = storedProducts.getStockMoveLine();
+      MassStockMove duplicateMassStockMove = storedProducts.getMassStockMove();
+      storedProducts.setCurrentQty(storedQty);
+      StoredProducts duplicateStoredProduct =
+          createStoredProduct(
+              duplicateProduct,
+              duplicateTrackingNumber,
+              currentQty.subtract(storedQty),
+              duplicateToStockLocation,
+              BigDecimal.ZERO,
+              duplicateStockMoveLine,
+              duplicateMassStockMove);
+      storeProductsRepository.save(duplicateStoredProduct);
+    }
+
+    if (storedProducts.getStoredQty().compareTo(BigDecimal.ZERO) != 0
+        && storedProducts.getToStockLocation() != null
+        && storedProducts.getStockMoveLine() == null) {
+
+      if (storedProducts.getStoredQty().compareTo(storedProducts.getCurrentQty()) == 1
+          && storedProducts.getTrackingNumber() != null) {
+        throw new AxelorException(
+            0,
+            I18n.get(StockExceptionMessage.STORED_QUANTITY_GREATER_THAN_CURRENT_QTY)
```

Comment:
We can use format instead of using operator '+',the message can be for example "Stored quantity is greater than current quantity (%s %s)"